Pokemon expansion debug

TheXaman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Adds a debug screen for pokemon front, back and icon sprites, including animations, gender differences, shiny palettes and mon cries. Its accessed via the pokemon summary screen and pressing the SELECT button. The debug define is P_ENABLE_DEBUG in /include/pokemon_debug.h .

## **Discord contact info**
<!--- formatted as name#numbers, e.g. PikalaxALT#5823 -->
<!--- Contributors must join https://discord.gg/d5dubZ3 -->
TheXaman#5612

AsparagusEduardo commented on include/pokemon_debug.h:
```diff
@@ -0,0 +1,9 @@
+#ifndef GUARD_POKEMON_DEBUG_H
+#define GUARD_POKEMON_DEBUG_H
+
+#define P_ENABLE_DEBUG
```
`P_ENABLE_DEBUG` should be moved to `pokemon_config.h`

TheXaman commented:
Changed the location like requested to `pokemon_config.h`.

LOuroboros commented on include/constants/pokemon_config.h:
```diff
@@ -18,6 +18,8 @@
 #define P_UPDATED_EGG_GROUPS    GEN_8 // Since Gen 8, certain Pokémon have gained new egg groups.
 #define P_SHEDINJA_BALL         GEN_8 // Since Gen 4, Shedinja requires a Poké Ball for its evolution. In Gen 3, Shedinja inherits Nincada's Ball.
 
+#define P_ENABLE_DEBUG                // Enables a debug menu for pokemon sprites and icons, accessed by pressing SELECT in the summary screen.
+
 #ifndef ITEM_EXPANSION
```
```suggestion
#define P_SHEDINJA_BALL         GEN_8 // Since Gen 4, Shedinja requires a Poké Ball for its evolution. In Gen 3, Shedinja inherits Nincada's Ball.

#define P_ENABLE_DEBUG           TRUE // If set to TRUE, enables a debug menu to use for pokemon sprites and icons in the summary screen accessed by pressing SELECT.
```

To remain consistent with the BE's.
For maximum consistency, it should be moved to the end of the file, just before the guard constants.

AsparagusEduardo replied:
I feel the one in BE should be moved up. It's a pretty big feature that some people might miss because it's so far down.

LOuroboros replied:
That works too :eyes:

LOuroboros commented on src/decompress.c:
```diff
@@ -113,6 +114,45 @@ void LoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32
     DrawSpindaSpots(species, personality, dest, isFrontPic);
 }
 
+#ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -22,6 +23,10 @@ struct MonIconSpriteTemplate
 
 // static functions
 static u8 CreateMonIconSprite(struct MonIconSpriteTemplate *, s16, s16, u8);
+#ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2575,6 +2580,33 @@ u8 CreateMonIcon(u16 species, void (*callback)(struct Sprite *), s16 x, s16 y, u
     return spriteId;
 }
 
+#ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2646,6 +2678,13 @@ const u8 *GetMonIconPtr(u16 species, u32 personality)
     return GetMonIconTiles(GetIconSpecies(species, personality), personality);
 }
 
+#ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2715,6 +2754,18 @@ const u8* GetMonIconTiles(u16 species, u32 personality)
     return iconSprite;
 }
 
+#ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_summary_screen.c:
```diff
@@ -1555,6 +1556,15 @@ static void Task_HandleInput(u8 taskId)
             PlaySE(SE_SELECT);
             BeginCloseSummaryScreen(taskId);
         }
+        #ifdef P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/decompress.c:
```diff
@@ -113,6 +114,45 @@ void LoadSpecialPokePic(const struct CompressedSpriteSheet *src, void *dest, s32
     DrawSpindaSpots(species, personality, dest, isFrontPic);
 }
 
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -22,6 +23,10 @@ struct MonIconSpriteTemplate
 
 // static functions
 static u8 CreateMonIconSprite(struct MonIconSpriteTemplate *, s16, s16, u8);
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2575,6 +2580,33 @@ u8 CreateMonIcon(u16 species, void (*callback)(struct Sprite *), s16 x, s16 y, u
     return spriteId;
 }
 
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2646,6 +2678,13 @@ const u8 *GetMonIconPtr(u16 species, u32 personality)
     return GetMonIconTiles(GetIconSpecies(species, personality), personality);
 }
 
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_icon.c:
```diff
@@ -2715,6 +2754,18 @@ const u8* GetMonIconTiles(u16 species, u32 personality)
     return iconSprite;
 }
 
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_summary_screen.c:
```diff
@@ -1555,6 +1556,15 @@ static void Task_HandleInput(u8 taskId)
             PlaySE(SE_SELECT);
             BeginCloseSummaryScreen(taskId);
         }
+        #if P_ENABLE_DEBUG
```
```suggestion
        #if P_ENABLE_DEBUG == TRUE
```

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -0,0 +1,754 @@
+    LoadPalette(gDecompressionBuffer, paletteOffset, 0x20);
+    LoadPalette(gDecompressionBuffer, 0x80 + battlerId * 16, 0x20);
+
+    //if (species == SPECIES_CASTFORM)
+    //{
+    //    paletteOffset = 0x100 + battlerId * 16;
+    //    LZDecompressWram(lzPaletteData, gBattleStruct->castformPalette[0]);
+    //    LoadPalette(gBattleStruct->castformPalette[gBattleMonForms[battlerId]], paletteOffset, 0x20);
+    //}
+
+    // transform's pink color
+    //if (gBattleSpritesDataPtr->battlerData[battlerId].transformSpecies != SPECIES_NONE)
+    //{
+    //    BlendPalette(paletteOffset, 16, 6, RGB_WHITE);
+    //    CpuCopy32(gPlttBufferFaded + paletteOffset, gPlttBufferUnfaded + paletteOffset, 32);
+    //}
```
Is this needed? :eyes:

Not a fan of keeping useless code in.

TheXaman replied:
No, an oversight - now deleted

TheXaman commented:
Added improvements by ShinyDragonHunter and AsparagusEduardo, fixed the overlapping animations and palette related bugs

AsparagusEduardo commented:
Something in the last couple commits broke some icons.
Before and after:
![debug2](https://user-images.githubusercontent.com/2904965/144717772-a2d74f81-c8da-406e-9d5e-ae1972786160.gif) ![debug1](https://user-images.githubusercontent.com/2904965/144717783-7e710355-1963-482b-9267-e91123bd2253.gif)

ShinyDragonHunter commented:
> Something in the last couple commits broke some icons. Before and after:
> 
> EDIT: The icons work fine when reading from party menu at least
I think it's because the front pic now uses the same slot as the mon icon palette slot

AsparagusEduardo commented on src/pokemon_debug.c:
```diff
@@ -0,0 +1,740 @@
+#include "constants/items.h"
+
+//Defines
+#define DEBUG_MON_X 137
+#define DEBUG_MON_Y 18
+#define DEBUG_MON_BACK_X 32
```
I messed up and accidentally moved the coords in the wrong direction when submitting the changes, but this should fix it. I also fully tested what X value for the backsprite is needed for syncing the back sprite to how it is in battle.
```suggestion
#define DEBUG_MON_X 144
#define DEBUG_MON_Y 11
#define DEBUG_MON_BACK_X 40
```

TheXaman commented:
> 



> Something in the last couple commits broke some icons. Before and after: ![debug2](https://user-images.githubusercontent.com/2904965/144717772-a2d74f81-c8da-406e-9d5e-ae1972786160.gif) ![debug1](https://user-images.githubusercontent.com/2904965/144717783-7e710355-1963-482b-9267-e91123bd2253.gif)
> 
> EDIT: The icons work fine when reading from party menu at least ![image](https://user-images.githubusercontent.com/2904965/144717915-b6bbd89e-c00d-4baa-9e39-274f12dfd95f.png)

fixed now as far as I can tell. Reason for the bug was: previously it loaded all possible icon palettes at once and overwrote one with the front sprite palette, resulting in some icons breaking if its palette was overwritten

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -40,13 +40,13 @@
 #define DEBUG_MON_BACK_Y 50
-#define DEBUG_ICON_X 158
-#define DEBUG_ICON_Y 90
+#define DEBUG_ICON_X 19//158 + 32
+#define DEBUG_ICON_Y 19//90 + 40
```
Are the comments necessary? :eyes:
They don't even seem to be accurate?

TheXaman replied:
no, those were the old coordinates for the icon in the bottom right - I'll remove them in the next commit

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -692,11 +824,18 @@ void CB2_Debug_Pokemon(void)
             FreeAllSpritePalettes();
             gReservedSpritePaletteCount = 8;
             ResetAllPicSprites();
+            //BlendPalettes(PALETTES_ALL, 16, RGB_BLACK);
```
Is this line that is commented out needed? :eyes:

TheXaman replied:
in this case yes for later

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -623,6 +656,107 @@ void BattleLoadOpponentMonSpriteGfxCustom(u16 species, bool8 isFemale, bool8 isS
 }
 
+//Battle background functions
+struct BattleBackground
+{
+    const void *tileset;
+    const void *tilemap;
+    const void *entryTileset;
+    const void *entryTilemap;
+    const void *palette;
+};
```
This should be put near the top of the file alongside the other structs such as `PokemonDebugModifyArrows` or `PokemonDebugMenu` imo.

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -623,6 +656,107 @@ void BattleLoadOpponentMonSpriteGfxCustom(u16 species, bool8 isFemale, bool8 isS
+    const void *palette;
+};
+extern const struct BattleBackground sBattleTerrainTable[];
+#define MAP_BATTLE_SCENE_NORMAL       0
+#define MAP_BATTLE_SCENE_GYM          1
+#define MAP_BATTLE_SCENE_MAGMA        2
+#define MAP_BATTLE_SCENE_AQUA         3
+#define MAP_BATTLE_SCENE_SIDNEY       4
+#define MAP_BATTLE_SCENE_PHOEBE       5
+#define MAP_BATTLE_SCENE_GLACIA       6
+#define MAP_BATTLE_SCENE_DRAKE        7
+#define MAP_BATTLE_SCENE_FRONTIER     8
+#define MAP_BATTLE_SCENE_LEADER       9
+#define MAP_BATTLE_SCENE_WALLACE      10
+#define MAP_BATTLE_SCENE_GROUDON      11
+#define MAP_BATTLE_SCENE_KYOGRE       12
+#define MAP_BATTLE_SCENE_RAYQUAZA     13
```
Not sure how I feel about these constants, or the ones near the top of the file.

They should all be put inside an `include/constants/pokemon_debug.h` file imo.

TheXaman replied:
followed your suggestion and added a `include/constants/pokemon_debug.h` file for defines

TheXaman commented:
**Changes:**

- [x]  added the option to view all front and back animations on every mon
- [x]  merged with upstream
- [x] added battle backgrounds and platforms and added the option to change them

**Missing:**
- [ ] new menu navigation to properly access all the new options
- [ ] add proper elevation/height calculation for front sprites (flying mons and their shadows)

LOuroboros commented on src/pokemon_debug.c:
```diff
@@ -623,6 +656,107 @@ void BattleLoadOpponentMonSpriteGfxCustom(u16 species, bool8 isFemale, bool8 isS
+    const void *entryTilemap;
+    const void *palette;
+};
+extern const struct BattleBackground sBattleTerrainTable[];
```
This should be defined somewhere near the top of the file, maybe below `sBgColor` imo.

TheXaman commented:
**Menu navigation suggestion**
Buttons that stay the same:
-L: play back anim (plays cry)
-R: play front anim (plays cry)
-START: shiny
-SELECT: gender difference
Main menu:
-A: enter submenu
-B: exit menu
-DPAD: change mon ID
Submenu:
-B: back to main menu
-DPAD LEFT and RIGHT: change value
-DPAD UP and DOWN: change option
    -option0: change form
    -option1: change back anim
    -option2: change front anim
    -option3: change battle background

TheXaman commented:
Changed menu layout based on my suggestion:
![pokeemerald-0](https://user-images.githubusercontent.com/48356183/145679460-73b43cd9-3645-4c60-9dd8-4234d763ab06.png)

LOuroboros commented on include/battle_bg.h:
```diff
@@ -1,6 +1,17 @@
 #ifndef GUARD_BATTLE_BG_H
 #define GUARD_BATTLE_BG_H
 
+#if P_ENABLE_DEBUG
```
```suggestion
#if P_ENABLE_DEBUG == TRUE
```

TheXaman commented:
**Additions:**

- Proper front sprite y coordinate calculation including picCoords and elevation
- Elevation based shadows for flying mons
- Added footprints window
- Possibility to change and view values for picCoords and elevation live

https://user-images.githubusercontent.com/48356183/145892689-6581502a-4731-4059-86ea-8a96ad75ca60.mp4

AsparagusEduardo reviewed (commented):
Here's some stuff I've noticed:

I feel that shadows should load dynamically depending if the elevation is larger than 0 or not, to show how elevation affect it compared to changing the Y coordinate.
![7pj39nDxJN](https://user-images.githubusercontent.com/2904965/145910977-af7c70c0-e5f9-44bb-9024-8d0bc506e75a.gif)

When switching forms, the front and back animations are not updated:
![fQ01F0TBQB](https://user-images.githubusercontent.com/2904965/145910877-e1dd05ae-976e-4a56-849e-13fcfa6171ff.gif)

Forms switching wraps when moving to the right, but not to the left:
![sK3da36ML5](https://user-images.githubusercontent.com/2904965/145911333-b1ab52ba-4b0a-4c08-9578-038b56530c25.gif)

AsparagusEduardo commented:
Another thing, the Groudon battle scene seems to be loading the Kyogre battle scene instead
![7EuDOJFdWQ](https://user-images.githubusercontent.com/2904965/145912878-1450f638-356a-4c5b-8b00-112a1ae06e66.gif)

TheXaman commented:
> Another thing, the Groudon battle scene seems to be loading the Kyogre battle scene instead ![7EuDOJFdWQ](https://user-images.githubusercontent.com/2904965/145912878-1450f638-356a-4c5b-8b00-112a1ae06e66.gif)

Approved your PR, fixed the Groudon terrain and made the shadow dynamically show up:

https://user-images.githubusercontent.com/48356183/145967503-47bc24d1-b945-46f8-bf15-80849f274198.mp4

MandL27 reviewed (approved):
lgtm
